HP-318 step 9: change language and improve accessibility

NikoHelle opened the pull request:
Language can be changed and saved to Profile API.

Added accessibility improvements:
- when a form is closed, focus is returned to the button that opened a form
- aria-labels of edit buttons have better description of the target
- focus is trapped inside a form until it is saved

Also deleted unused files

NikoHelle committed:
(HP-318) Added profile language selection

NikoHelle committed:
(HP-318) Accessibility: return focus to previous element

When dynamically changing UI, like showing a form and hiding it after save, the focus should return to the element that initiated the change, so keyboard users can continue from there.

Focus is now returned to edit or 'add new' button, when action is done or cancelled.

NikoHelle committed:
(HP-318) Accessible action labels

Buttons in EditButton component now have informative aria-labels that include the value the button targets. For example
"Remove/Edit email <the email>".

NikoHelle committed:
(HP-318) Accessibility improvment with focus keeper

When a form is rendered, focus should be kept inside the form until it is saved or cancelled. Keeping focus inside prevents screen reader users to get lost on the page and wander off the form.

NikoHelle committed:
(HP-318) Remove unused files

akikoskinen commented on src/common/focusKeeper/__tests__/FocusKeeper.test.tsx:
```diff
@@ -0,0 +1,47 @@
+      if (id === 'target') {
+        focusEventCount = focusEventCount + 1;
+      }
+    };
```
About names: instead of a general `focusListener` function name, this could have a more descriptive name, maybe something like `incrementTargetElementFocusCounter`. Similarly, since `focusEventCount` seems to count only `target` element's focuses, it could be called e.g. `targetElementFocusCount`.

NikoHelle replied:
Second listener was obsolete, so removed it. Kept same function name, because it is a focusListener (now for single element), but changed focusEventCount to focusEventsOnTarget

akikoskinen commented on src/common/focusKeeper/__tests__/FocusKeeper.test.tsx:
```diff
@@ -0,0 +1,47 @@
+      <div>
+        <FocusKeeper targetId={'target'} className="focus-keeper">
+          <input type="text" id="target" onFocus={focusListener} />
+          <input type="text" id="second" onFocus={focusListener} />
```
Since `focusListener` counts only events for the `target` element, it doesn't have to be added to the `second` element?

NikoHelle replied:
Removed

akikoskinen commented on src/common/focusKeeper/__tests__/FocusKeeper.test.tsx:
```diff
@@ -0,0 +1,47 @@
+      ) as HTMLElement;
+    };
+    const focusTrapperAbove = getTrapperElement('above');
+    const focusTrapperBelow = getTrapperElement('below');
```
IMO the "above"/"below" terminology is somewhat obscure. "Above" (or "below") what? The CSS selectors seem to provide better terms: "first child", "last child". That terminology could be used instead of "above" and "below".

NikoHelle replied:
Changed above -> before, below -> after and added better explanation which is which.

akikoskinen commented on src/common/focusKeeper/__tests__/FocusKeeper.test.tsx:
```diff
@@ -0,0 +1,47 @@
+    focusTrapperAbove.focus();
+    result.container.focus();
+    focusTrapperBelow.focus();
+    expect(focusEventCount).toEqual(2);
```
Should the focus count be checked after each step, to ensure that the automatic focus changing works in the expected way?

NikoHelle replied:
Thx! Added more checks

akikoskinen commented on src/common/focusKeeper/__tests__/FocusKeeper.test.tsx:
```diff
@@ -0,0 +1,47 @@
+    const focusTrapperAbove = getTrapperElement('above');
+    const focusTrapperBelow = getTrapperElement('below');
+    // move focus outside of FocusKeeper
+    focusTrapperAbove.focus();
```
This comment seems a bit misplaced, as the following line seems to do exactly the opposite, moving the focus **inside** the `FocusKeeper`. Perhaps there doesn't have to be a comment here at all.

NikoHelle replied:
Yeah, moved it down to correct line.

akikoskinen commented on src/profile/components/additionalInformation/AdditionalInformation.tsx:
```diff
@@ -0,0 +1,101 @@
+  const editData = editDataList[0];
+  const { value } = editData;
+  const { language } = value as AdditionalInformationValue;
+  const languageOptions: OptionType[] = profileConstants.LANGUAGES.map(
+    languageOption => ({
+      value: languageOption,
+      label: t(`LANGUAGE_OPTIONS.${languageOption}`),
+    })
+  );
```
Is this a static value? Would it make sense to construct it outside the component function and then just use it here?

NikoHelle replied:
It changes when UI language changes, so it is not static.

akikoskinen commented on src/profile/components/additionalInformation/AdditionalInformation.tsx:
```diff
@@ -0,0 +1,101 @@
+        <Formik
+          initialValues={{ language }}
+          onSubmit={() => {
+            /* onChange does this. This comment is here because onSubmit is mandatory and function cannot be empty */
```
Kind of weird, that every other change in the data requires a "start modifying/modify/save" cycle, but this one field is saved immediately :thinking:

akikoskinen replied:
Also, IMO this leads to a bit sub-par UX in error situations. With the current solution, when one selects another language, but the saving fails, the newly selected language remains selected, and an error is shown which asks to "try again". How does the user try again when there's no save button? The user has to temporarily select a third language, then reselect the one they originally wanted to select and hope that this time the saving succeeds.

I wonder if this is the best flow for the user...

NikoHelle replied:
Good point and clearly a bug. The selected value is now reverted when an error occurs, so change can be made again without doing any dummy changes.

IMO, this is much more convenient for the user than rendering first:

"Profile language: English [edit]"

And then

"Profile language: [English ˆ]"
[Save] [Cancel]

Someday when there is more Additional information than just language dropdown, edit mode could be used. 

Changing a primary / removing an item does not require edit mode either.

akikoskinen replied:
> Changing a primary / removing an item does not require edit mode either.

Very good point, somehow I didn't realize this earlier. This is good now.

akikoskinen commented on src/profile/components/additionalInformation/__tests__/AdditionalInformation.test.tsx:
```diff
@@ -0,0 +1,120 @@
+    return Promise.resolve(testTools);
+  };
+
+  it('should render language selection, auto save and show success/error message', async () => {
```
This one test has quite a many responsibilities. It seems that it could be split at least into two, the success and error cases.

akikoskinen commented on src/profile/helpers/createActionAriaLabels.ts:
```diff
@@ -0,0 +1,49 @@
+      )} ${value}`,
+      remove: `${t('profileForm.remove')} ${t(
+        'profileInformation.address'
+      )} ${value}`,
```
Data should be inserted into the translations with place holders, not by concatenating strings in code. Nor should translated strings be concatenated together to form new user visible strings.

NikoHelle committed:
(HP-318) Get latest item version in EditData

UI might pass old version of data item. Always get the latest version with item.id

akikoskinen commented on src/profile/helpers/createActionAriaLabels.ts:
```diff
@@ -0,0 +1,42 @@
+      edit: `${t(editTranslation, { value })}`,
+      primary: `${t(primaryTranslation)}`,
+      setPrimary: `${t(ariaSetPrimaryWithValue, { value })}`,
+      remove: `${t(ariaRemoveWithValue, { value })}`,
```
Do the `t` calls have to be within a string? Does `t` return something else than a string?

NikoHelle replied:
No, t returns a string. Fixed. 
Even if a translation is not found for a key, it still returns the key.

akikoskinen commented on src/i18n/fi.json:
```diff
@@ -138,7 +138,7 @@
     "primary": "Ensisijainen",
     "remove": "Poista",
-    "setPrimary": "Aseta ensisijaiseksi",
+    "setPrimary": "Aseta ensisijaiseksi ",
```
What's with the extra space?

NikoHelle replied:
From the translation Excel. Removed it

akikoskinen commented on src/i18n/fi.json:
```diff
@@ -184,6 +184,12 @@
+    "ariaSetPrimaryEmailWithValue": "Aseta {{value}} ensisijaiseksi sähköpostiksi",
+    "ariaSetPrimaryPhoneWithValue": "Aseta {{value}} ensisijaiseksi puhelinnumeroksi",
+    "ariaRemoveAddressWithValue": "Poista osoite {{value}} ",
+    "ariaRemoveEmailWithValue": "Poista sähköposti {{value}}",
```
Pitäisikö käyttää pitempää "sähköpostiosoite" termiä, niin kuin näyttäisi, että on muissakin kohdissa käytetty?

NikoHelle replied:
Joo, laitoin saman.

akikoskinen commented on src/i18n/en.json:
```diff
@@ -184,6 +184,12 @@
+    "ariaSetPrimaryPhoneWithValue": "Set {{value}} as primary phone",
+    "ariaRemoveAddressWithValue": "Remove address {{value}}",
+    "ariaRemoveEmailWithValue": "Remove email {{value}}",
+    "ariaRemovePhoneWithValue": "Remove phone {{value}}",
```
Should the term "phone number" be used instead of "phone"? Both terms seem to be in use, but feels like "phone number" better refers to the _number_, whereas just "phone" refers to the device.

NikoHelle replied:
Both seems to be used. I changed new and old ones to use "phone number"

akikoskinen reviewed (commented):
A few more comments, nothing biggy.

akikoskinen reviewed (approved):
:+1:
